DOC-12485 prevent bucket from running out of space (#3811)
couchbase/docs-server: modules/learn/pages/buckets-memory-and-storage/storage-settings.adoc

[#understanding-couchbase-storage]
== Understanding Couchbase Storage

In addition to storing data in memory, Couchbase Server also stores data in Couchbase buckets on disk.
Saving data to disk provides persistence so that data is not lost if a node restarts or fails.
It also lets your data sets exceed the limits of your available memory.
Couchbase Server restores data that's not in memory from disk when needed.

Ephemeral buckets and their items exist only in memory and are never written to disk.
For more details, see xref:buckets-memory-and-storage/buckets.adoc[Buckets].

Couchbase Server compresses the data it writes to disk.
Compression reduces the amount of disk space used which can help reduce costs.
It also makes the backup and restore procedures easier.
In addition to compressing data written to disk, Couchbase Server can also compress data in memory.
See xref:buckets-memory-and-storage/compression.adoc[Compression] for more information.

> Disk access does not interrupt most client interactions.

Reviewer: "Disk access does not interrupt most client interactions." We should probably get rid of this line. Durable writes which are client operation require flush to disk.

[#understanding-couchbase-storage]
== Understanding Couchbase Storage

In addition to storing data in memory, Couchbase Server also stores data in Couchbase buckets on disk.
Saving data to disk provides persistence so that data is not lost if a node restarts or fails.
It also lets your data sets exceed the limits of the memory in your cluster.
Couchbase Server restores data that's not in memory from disk when needed.

Ephemeral buckets and their items exist only in memory and are never written to disk.
For more details, see xref:buckets-memory-and-storage/buckets.adoc[Buckets].

Couchbase Server compresses the data it writes to disk.
Compression reduces the amount of disk space used which can help reduce costs.

> It also makes the backup and restore procedures easier.

Reviewer 2: I'm not sure what compression has to do with the ease of backup and restore. Does this just mean speed/performance?

Author: I'm not sure either. I've removed it.

In addition to compressing data written to disk, Couchbase Server can also compress data in memory.
See xref:buckets-memory-and-storage/compression.adoc[Compression] for more information.

Disk access does not interrupt most client interactions.
However, if a client requests an item that's on disk, it must wait while Couchbase Server reads, decompresses, and copies the data into memory. 

You can remove items from disk based on a configured expiration time, called time to live.
See xref:data/expiration.adoc[Expiration] for details.

> To see how Couchbase Server saves new items and updates existing items in Couchbase buckets, using both memory and storage, seexref:buckets-memory-and-storage/memory-and-storage.adoc[Memory and Storage].

Reviewer 2: ```suggestion
To see how Couchbase Server saves new items and updates existing items in Couchbase buckets, using both memory and storage, see xref:buckets-memory-and-storage/memory-and-storage.adoc[Memory and Storage].
```

[#threading]
== Threading

Couchbase Server uses synchronized, multi-threaded readers and writers to provide high-performance, simultaneous operations for data on disk.
Readers and writers each have their own set of threads.
To prevent conflicts, each thread is responsible for reading or writing a subset of the vBuckets in a Couchbase bucket.

You can control the number of reader and writer threads. 
In the Couchbase Server Web Console, you can have Couchbase Server automatically choose a default value or a value that optimizes disk I/O. 
You can also manually set the number of threads per node to a value between 1 and 64.

> Using a higher number of threads may improve performance if your hardware supports it, such as when your CPU has a high large of cores.

Reviewer 2: ```suggestion
Using a higher number of threads may improve performance if your hardware supports it, such as when your CPU has a larger number of cores.
```

Increasing the number of writer threads helps optimize durable writes.

> For more information, see xref:learn:data/durability.adoc[Durability].

Reviewer 2: This feels like it applies to the whole paragraph, when it really is just relevant to the prior sentence, perhaps it could be phrased better?

Author: I broke the paragraph into two to make the reference clearer.

You can specify the location where Couchbase Server stores data on a node for the following services:

* Data Service
* Index Service
* Analytics Service
* Eventing Service

In addition, you can use local paths for backup repositories.
See xref:learn:services-and-indexes/services/backup-service.adoc#repositories[Repositories] for more information.

Couchbase Server has a default storage location for logs that's platform-specific. 
For example, on Linux, the default location is `/opt/couchbase/var/lib/couchbase/logs`.

For information about setting data paths, see xref:manage:manage-nodes/initialize-node.adoc[Initialize a Node].

[[filesystem-free-space-and-usage-limits]]
=== Filesystem Free Space and Usage Limits

Running out of disk space on any filesystem can cause errors. 
In particular, running out of disk space on the filesystem containing the Data Service storage path can make recovery difficult.
Recovery problems could lead to data loss.

By default, Couchbase Server alerts you if the filesystem containing the following storage paths becomes more than 90% full:

* Data Service
* Index Service

> * `ns_log` and `audit_log` paths`

Reviewer 3: This comment is for lines 128 through 132:
By default, Couchbase Server alerts you if the filesystem containing the Data Service storage path becomes more than 75% full (10% before the limit, which is set to 85% by default).

The Index Service on line 131 and `ns_log` and `audit_log` paths on line 132 should be removed.

The maxDataDiskUsedPerc only applies to the data service storage path. So, the sentence should just say the data service storage path. If you don't specify a custom path for data, indexes, eventing, analytics when you initialize a cluster or a node, then, it's not easy to tell what goes into each individual storage paths since they then all end up in a default location. The ns_log does not get put in the data service storage path. Not sure what audit_log is ... but if it's the audit.log, then, that gets put into a location specified by the customer, so it shouldn't be mentioned here either.

I think that you are getting maxDiskUsedPerc (value 90) and maxDataDiskUsedPerc (value 75) mixed up -- on line 134 below, you have maxDiskUsedPerc when you mean to have maxDataDiskUsedPerc. 

cc @Peter-Searby

See xref:manage:manage-settings/configure-alerts.adoc[] for more information about alerts.

> You can change how full the disk becomes before triggering this alert by changing the  xref:rest-api:rest-cluster-email-notifications.adoc#maxdatadiskusedperc[maxDiskUsedPerc] alert limit.

Reviewer 3: On line 135, you have maxDiskUsedPerc showing in the text when you mean maxDataDiskUsedPerc 

% curl -s -X GET http://localhost:8091/settings/alerts/limits -u Administrator:password | jq .
{
  "certExpirationDays": 30,
  "historyWarningThreshold": 90,
  "lowIndexerResidentPerc": 10,
  "maxDataDiskUsedPerc": 75, <---- this is one you want
  "maxDiskUsedPerc": 90,   <---- getting it mixed up with this
  "maxIndexerRamPerc": 75,
  "maxOverheadPerc": 50,
  "memcachedSystemConnectionWarningThreshold": 90,
  "memcachedUserConnectionWarningThreshold": 90,
  "memoryCriticalThreshold": 90,
  "memoryNoticeThreshold": -1,
  "memoryWarningThreshold": 85,
  "stuckRebalanceThresholdIndex": 1800,
  "stuckRebalanceThresholdKV": 1800
}

Beyond alerting you, Couchbase Server does not take any action to prevent the filesystem from becoming full.

> You can enable a feature to have Couchbase Server stop writing to the Data Service storage path when it reaches a certain percentage of disk usage.

Reviewer 2: ```suggestion
You can enable a feature to have Couchbase Server Data Service stop writing to the Data Service storage path when it reaches a certain percentage of disk usage.
```
Also worth noting that this storage path may be on the same disk as other data, which may still be written to

This approach helps maintain file consistency and reduces the risk of file corruption.
Every time you add, modify, or delete data, Couchbase Server creates a new entry at the end of the data files.
As a result, files grow in size even when you delete data.

To prevent data files from growing too large, Couchbase Server periodically compacts them.
Compaction rewrites the file, applying additions, modifications, and deletions before saving a new version of the file.
You can change the schedule Couchbase Server follows to compact data.
See xref:manage:manage-settings/configure-compact-settings.asdoc[Auto-Compaction] for more information.
For information about configuring auto-compaction with the command line, see xref:cli:cbcli/couchbase-cli-setting-compaction.adoc[setting-compaction].

You can also perform compaction manually on a specific bucket.
For information about performing manual compaction with the command line, see xref:cli:cbcli/couchbase-cli-bucket-compact.adoc[bucket-compact].

For all information about using the REST API for compaction, see the xref:rest-api:compaction-rest-api.adoc[Compaction API].

> == Disk I/O Priority

Reviewer 2: Assuming this is the Bucket Priority (in the UI), my understanding is that this doesn't actually do anything. I'm not too sure why we've kept the config around, but it would be worth getting confirmation from KV how this should be documented (they look to be planning on cleaning this up in Ponyo: https://jira.issues.couchbase.com/browse/MB-66579)

Reviewer 3: @owend74 Do we need to change anything (from what is in the current documentation for "Disk I/O Priority" for 8.0 documentation (in light of MB-66579)?  Since MB-66579 is still open, I'm thinking that we should just leave things the way it is for now and clean-up the item from documentation and the UI in Totoro or Ponyo (based on how MB-66579 resolves).  Please advise.

Reviewer 3: FYI -- there's a new Doc ticket opened for the Bucket Priority issue --  [DOC-13332](https://jira.issues.couchbase.com/browse/DOC-13332)

from Dan Owen's email:

I chatted to Vesko Karaganev and confirmed in Morpheus Disk I/O Priority  (aka Bucket Priority) does not do anything, (and has not since 7.0.0)
 
_“It used to affect scheduling before we moved to the folly scheduler.”_

- See [MB-36956](https://jira.issues.couchbase.com/browse/MB-36956)

 
It is important we do NOT break backward compatibility, so we don’t want to make any interface changes at KV level and REST API.  However, I have
 
- Raised [MB-67281](https://jira.issues.couchbase.com/browse/MB-67281) - a CLI ticket to add a message.  The commands will be executed as normal, but a message will be displayed stating that bucket-priority has no effect.
- Raised [MB-67280](https://jira.issues.couchbase.com/browse/MB-67280) - a UI ticket to remove the option from the UI (release note request, associated with ticket)
- Raised [DOC-13332](https://jira.issues.couchbase.com/browse/DOC-13332) – to track the associated documentation changes.

Author: I actually stumbled across those isuees while doing the bucket eviction feature doc.  I did miss removing the Disk I/O Priority removal in that PR, though. I'm adding a comment to myself to fix it there, because I made a bunch of changes in that PR that are closer to the Disk I/O section in the page, and I'd like to avoid giving git any reason to create a merge conflict between these two PRs.

* Data Service
* Index Service
* Analytics Service
* Eventing Service

In addition, you can use local paths for backup repositories.
See xref:learn:services-and-indexes/services/backup-service.adoc#repositories[Repositories] for more information.

Couchbase Server has a default storage location for logs that's platform-specific. 
For example, on Linux, the default location is `/opt/couchbase/var/lib/couchbase/logs`.

For information about setting data paths, see xref:manage:manage-nodes/initialize-node.adoc[Initialize a Node].

[[filesystem-free-space-and-usage-limits]]
=== Filesystem Free Space and Usage Limits

Running out of disk space on any filesystem can cause errors. 
In particular, running out of disk space on the filesystem containing the Data Service storage path can make recovery difficult.
Recovery problems could lead to data loss.

By default, Couchbase Server alerts you if the filesystem containing the data service's storage path becomes 75% full.
It begin alerting you when disk containing this path reaches within 10% of the threshold set by the `maxDataDiskUsedPerc` setting, which defaults to 85%. 

> See xref:manage:manage-settings/configure-alerts.adoc[] for more information about alerts.

Reviewer 4: URL text "Alerts" is missing from:
See xref:manage:manage-settings/configure-alerts.adoc[] for more information about alerts.

You can also have the Data Service stop writing to its storage path when it reaches a percentage of disk usage.
The default (and recommended) limit is 85% full which means the Data Service stops writing data if the filesystem is 85% or more full. 
Enabling this limit helps avoid potential issues with recovery.

When you set a data disk usage limit, Couchbase Server starts alerting you when the filesystem fills to within 10% of the threshold you set. 
For example, if you set the limit of 80%, Couchbase Server alerts you when the filesystem reaches 70% full.
This threshold overrides the `maxDataDiskUsedPerc` alert limit.

When the filesystem reaches the disk use limit, the Data Service stops writing to the Data Service storage path.
Any attempts to write to the Data Service storage path results in an `EBucketDiskSpace` error. 
To re-enable writes, reduce the disk use on the filesystem to be less than the limit you set.

To learn how to set the disk usage limit using the Couchbase Server Web Console, see xref:manage:manage-settings/general-settings.adoc#data-settings[Data Settings].

> To set the limits using the REST API, see xref:rest-api:disk-usage-limits.adoc[].

Reviewer 4: URL text "Set Data Disk Use Limits" is missing from:
To set the limits using the REST API, see xref:rest-api:disk-usage-limits.adoc[].

A tombstone records an item removed from the database.
Couchbase Server uses tombstones to maintain consistency between nodes and clusters.

> ed to provide eventual consistency, between nodes and between clusters.

Reviewer 2: Who's ed?
